src/ray/util/file_persistence.cc

```diff
@@ -0,0 +1,75 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/file_persistence.h"
+
+#include <chrono>
+#include <filesystem>
+#include <fstream>
+#include <string>
+#include <thread>
+
+namespace ray {
+
+Status WriteFile(const std::string &file_path, const std::string &value) {
+  std::string tmp_path = file_path + ".tmp";
+  std::ofstream file(tmp_path);
+  if (!file.is_open()) {
+    return Status::IOError("Failed to open temp file for writing: " + tmp_path);
+  }
+  file << value;
+  file.close();
+  if (file.fail()) {
+    return Status::IOError("Failed to write to temp file: " + tmp_path);
+  }
+
+  std::error_code ec;
+  // On Windows, rename doesn't overwrite existing files, so remove first.
+  std::filesystem::remove(file_path, ec);
+  std::filesystem::rename(tmp_path, file_path, ec);
+  if (ec) {
+    return Status::IOError("Failed to rename " + tmp_path + " to " + file_path + ": " +
+                           ec.message());
+  }
+  return Status::OK();
+}
+
+StatusOr<std::string> WaitForFile(const std::string &file_path,
```

Comment:
So, one new thing we're trying is to use StatusSet.  StatusSet makes it a compile time check that the Status error types returned from this function are accounted for.  Take a look here for an example for how to use it: https://github.com/ray-project/ray/blob/2fa4348b658f8164ee00bef24b177a4a53717cc4/src/ray/core_worker/core_worker.cc#L1352